File: 1171-remove-zero-sum-consecutive-nodes-from-linked-list/1171-remove-zero-sum-consecutive-nodes-from-linked-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def removeZeroSumSublists(self, head: Optional[ListNode]) -> Optional[ListNode]:
        con_sum = 0
        dummy = ListNode(0)
        dummy.next = head
        mp = {}
        mp[0] = dummy

        while head:
            con_sum += head.val
            if con_sum in mp:
                start = mp[con_sum]
                temp = start
                p_sum = con_sum
                while temp != head:
                    temp = temp.next
                    p_sum += temp.val
                    if temp != head:
                        del mp[p_sum]
                start.next = head.next
            else:
                mp[con_sum] = head
            head = head.next
        return dummy.next

File: LeetCode/1171-remove-zero-sum-consecutive-nodes-from-linked-list/1171-remove-zero-sum-consecutive-nodes-from-linked-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode removeZeroSumSublists(ListNode head) {
        ListNode dummy = new ListNode(0);
        dummy.next = head;
        Map<Integer, ListNode> mp = new HashMap<>();
        mp.put(0, dummy);
        int sum = 0;

        while (head != null) {
            sum += head.val;
            if (mp.containsKey(sum)) {
                int pSum = sum;
                ListNode start = mp.get(sum);
                ListNode temp  = start;
                while (temp != head) {
                    temp = temp.next;
                    pSum += temp.val;
                    if (temp != head) {
                        mp.remove(pSum);
                    }
                }
                start.next = head.next;
            } else {
                mp.put(sum, head);
            }
            head = head.next;
        }
        return dummy.next;
    }
}